app/models/immunisation_import.rb

```diff
@@ -74,6 +77,7 @@ def process!(patient_session:)
         .uniq(&:urn)
         .reject(&:invalid?)
         .map(&:attributes)
+        .map { _1.merge(imported_from: self) }
```

Comment:
Why not create locations more like the vaccination_records below? That seems less surprising to me than instantiating a `Location` object, converting it to attributes, and then creating a new `Location` object from that.

Reply:
Then again creating a Location object only to throw it away and do it again later probably isn't fast either so anyway, operational reasoning...

I can switch to `.map(&:save!)` or do assignments more in line with the way we do the records.

Reply:
Just had a look at how `create!` this is implemented, and I'm not sure there's any magic here that speeds it up, especially since it uses `save!` instead of `create!` when the input isn't an array:

```ruby
      def create!(attributes = nil, &block)
        if attributes.is_a?(Array)
          attributes.collect { |attr| create!(attr, &block) }
        else
          object = new(attributes, &block)
          object.save!
          object
        end
      end
```

https://github.com/rails/rails/blob/36c1591bcb5e0ee3084759c7f42a706fe5bb7ca7/activerecord/lib/active_record/persistence.rb#L50